src/app/InteractionModelEngine.h

```diff
@@ -204,7 +216,26 @@ void DispatchSingleClusterCommand(chip::ClusterId aClusterId, chip::CommandId aC
  * otherwise.
  */
 bool ServerClusterCommandExists(chip::ClusterId aClusterId, chip::CommandId aCommandId, chip::EndpointId aEndPointId);
-CHIP_ERROR ReadSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVWriter & aWriter);
+
+/**
+ *  Fetch attribute value and version info and write to the TLVWriter provided.
+ *  When the endpoint / cluster / attribute / event data specified by aClusterInfo does not exist, corresponding interaction model
+ * error code will be put into the writer, and CHIP_NO_ERROR will be returned and apDataExists will be set to false.
+ *  If the data exists on the server, the data (with tag kCsTag_Data) and the data version (with tag kCsTag_DataVersion) will be put
+ * into the TLVWriter and apDataExists will be set to true. TLVWriter error will be returned if any error occurred during encoding
+ * these values.
+ *  This function is implemented by CHIP as a part of cluster data storage & management.
+ * The apWriter and apDataExists can be nullptr.
+ *
+ *  @param[in]    aClusterInfo      The cluster info object, for the path of cluster data.
+ *  @param[in]    apWriter          The TLVWriter for holding cluster data. Can be a nullptr if the caller does not care
```

Comment:
This is still not making sense to me.  If the caller does not care about the value, what do they care about?  Just whether it exists?

But looking at the `ReadSingleClusterData` implemetations:

* The one in `TestCommandInteraction.cpp` leaves `apDataExists` unset if `apWriter` is null.
* The one in `TestReportingEngine.cpp`, same thing.
* The one in `chip_im_initiator.cpp`, same thing.
* The one in `chip_im_responder.cpp`, same thing.
* The one in `ember-compatibility-functions.cpp` seems to implement the "just set apDataExists" semantics, ok.

And more to the point: is there ever a caller that actually passes null for apWriter?  I don't obviously see one.

Reply:
https://github.com/project-chip/connectedhomeip/issues/7965